app/src/main/java/com/teamagam/gimelgimel/app/control/sensors/LocationFetcher.java

```diff
@@ -0,0 +1,183 @@
+package com.teamagam.gimelgimel.app.control.sensors;
+
+import android.Manifest;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.location.Location;
+import android.location.LocationListener;
+import android.location.LocationManager;
+import android.os.Bundle;
+import android.support.annotation.StringDef;
+import android.support.v4.app.ActivityCompat;
+
+import com.teamagam.gimelgimel.app.model.entities.LocationSample;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.util.ArrayList;
+import java.util.Collection;
+
+/**
+ * Handles location fetching against the system's sensors
+ */
+public class LocationFetcher {
+
+    @StringDef({
+            ProviderType.LOCATION_PROVIDER_GPS,
+            ProviderType.LOCATION_PROVIDER_NETWORK,
+            ProviderType.LOCATION_PROVIDER_PASSIVE
+    })
+    @Retention(RetentionPolicy.SOURCE)
+    public @interface ProviderType {
+        String LOCATION_PROVIDER_GPS = LocationManager.GPS_PROVIDER;
+        String LOCATION_PROVIDER_NETWORK = LocationManager.NETWORK_PROVIDER;
+        String LOCATION_PROVIDER_PASSIVE = LocationManager.PASSIVE_PROVIDER;
+    }
+
+
+    private static void checkForLocationPermission(Context context) {
```

Comment:
consider usins context only for checking permission once in the constructor and do not hold field with pointer to the context.